src/actix/certificate_helpers.rs

```diff
@@ -1,126 +1,168 @@
 use parking_lot::RwLock;
+use rustls::server::{AllowAnyAuthenticatedClient, ClientHello, ResolvesServerCert};
+use rustls::sign::CertifiedKey;
+use rustls::{Certificate, RootCertStore, ServerConfig};
+use rustls_pemfile::Item;
 
+use super::with_buf_read;
 use crate::settings::{Settings, TlsConfig};
 
-struct SslContextHolder {
-    ssl_context: SslContext,
+/// A TTL based rotating server certificate resolver
+struct RotatingCertificateResolver {
+    /// TLS configuration used for loading/refreshing certified key
     tls_config: TlsConfig,
-    verify_client_cert: bool,
-    refresh_interval: Duration,
-    last_updated: Instant,
+
+    /// TTL for each rotation
+    ttl: Option<Duration>,
+
+    /// Current certified key
+    key: RwLock<CertifiedKeyWithAge>,
 }
 
-impl SslContextHolder {
-    fn new(
-        tls_config: &TlsConfig,
-        verify_client_cert: bool,
-        refresh_interval: Duration,
-    ) -> Result<SslContextHolder, ErrorStack> {
-        let ssl_context = build_ssl_context(tls_config, verify_client_cert)?;
-        Ok(SslContextHolder {
-            ssl_context,
-            tls_config: tls_config.clone(),
-            verify_client_cert,
-            refresh_interval,
-            last_updated: Instant::now(),
+impl RotatingCertificateResolver {
+    pub fn new(tls_config: TlsConfig, ttl: Option<Duration>) -> io::Result<Self> {
+        let certified_key = load_certified_key(&tls_config)?;
+
+        Ok(Self {
+            tls_config,
+            ttl,
+            key: RwLock::new(CertifiedKeyWithAge::from(certified_key)),
         })
     }
 
-    fn try_get_ssl_context(&self) -> Option<&SslContext> {
-        (self.last_updated.elapsed() < self.refresh_interval).then_some(&self.ssl_context)
+    /// Get certificate key or refresh
+    ///
+    /// The key is automatically refreshed when the TTL is reached.
+    /// If refreshing fails, an error is logged and the old key is persisted.
+    fn get_key_or_refresh(&self) -> Option<Arc<CertifiedKey>> {
+        // Return current key if TTL has not elapsed
+        let mut key = self.key.read().key.clone();
+        match self.ttl {
+            None => return Some(key),
+            Some(ttl) if !self.key.read().is_expired(ttl) => return Some(key),
+            Some(_) => {}
+        }
+
+        // Refresh key
+        match self.refresh() {
+            Ok(certified_key) => key = certified_key,
+            Err(err) => log::error!("Failed to refresh TLS certificate, keeping current: {err}"),
+        }
+        Some(key)
+    }
+
+    /// Refresh certificate key, update and return
+    fn refresh(&self) -> io::Result<Arc<CertifiedKey>> {
+        log::info!(
+            "Refreshing TLS certificates for actix (TTL: {})",
+            self.ttl.map(|ttl| ttl.as_secs()).unwrap_or(0),
+        );
+
+        let mut lock = self.key.write();
+        let key = load_certified_key(&self.tls_config)?;
+        lock.replace(key.clone());
+        Ok(key)
     }
```

Comment:
Slightly better way to refresh the key (also see the following comments/suggestions):

```suggestion
    fn get_key_or_refresh(&self) -> Arc<CertifiedKey> {
        // Get read-only lock to the key. If TTL is not configured or is not expired, return key.
        let ttl = {
            let key = self.key.read();

             match self.ttl {
                Some(ttl) if key.is_expired(ttl) => ttl,
                _ => return key.key.clone(),
            }
        };

        // If TTL is expired:
        // - get read-write lock to the key
        // - *re-check that TTL is expired* (to avoid refreshing the key multiple times from concurrent threads)
        // - refresh and return the key
        let mut key = self.key.write();

        if key.is_expired(ttl) {
            if let Err(err) = key.refresh(&self.tls_config) {
                log::error!("Failed to refresh TLS certificate, keeping current: {err}");
            }
        }

        key.key.clone()
    }
```

Reply:
Good suggestions, thanks! I like the second TTL check to prevent the datarace.